include/internal/quic_wire.h

```diff
@@ -0,0 +1,690 @@
+    (((x) & ~(uint64_t)1) == OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_TRANSPORT)
+
+/*
+ * QUIC Frame Logical Representations
+ * ==================================
+ */
+
+/* QUIC Frame: ACK */
+typedef struct ossl_quic_ack_range_st {
+    /*
+     * Represents an inclusive range of packet numbers [start, end].
+     * start must be <= end.
+     */
+    QUIC_PN start, end;
+} OSSL_QUIC_ACK_RANGE;
+
+typedef struct ossl_quic_frame_ack_st {
+    /*
+     * A sequence of packet number ranges [[start, end]...].
+     *
+     * The ranges must be sorted in descending order, for example:
+     *      [ 95, 100]
+     *      [ 90,  92]
+     *      etc.
+     *
+     * As such, ack_ranges[0].end is always the highest packet number
+     * being acknowledged and ack_ranges[num_ack_ranges-1].start is
+     * aalways the lowest packet number being acknowledged.
+     *
+     * num_ack_ranges must be greater than zero, as an ACK frame must
+     * acknowledge at least one packet number.
+     */
+    OSSL_QUIC_ACK_RANGE        *ack_ranges;
+    size_t                      num_ack_ranges;
+
+    OSSL_TIME                   delay_time;
+    uint64_t                    ect0, ect1, ecnce;
+    char                        ecn_present;
+} OSSL_QUIC_FRAME_ACK;
+
+/* QUIC Frame: STREAM */
+typedef struct ossl_quic_frame_stream_st {
+    uint64_t                stream_id;  /* Stream ID */
+    uint64_t                offset;     /* Logical offset in stream */
+    uint64_t                len;        /* Length of data in bytes */
+    const unsigned char    *data;
+
+    /*
+     * On encode, this determines whether the len field should be encoded or
+     * not. If zero, the len field is not encoded and it is assumed the frame
+     * runs to the end of the packet.
+     *
+     * On decode, this determines whether the frame had an explicitly encoded
+     * length. If not set, the frame runs to the end of the packet and len has
+     * been set accordingly.
+     */
+    char                    has_explicit_len;
+
+    /* 1 if this is the end of the stream */
+    char                    is_fin;
+} OSSL_QUIC_FRAME_STREAM;
+
+/* QUIC Frame: CRYPTO */
+typedef struct ossl_quic_frame_crypto_st {
+    uint64_t                offset; /* Logical offset in stream */
+    uint64_t                len;    /* Length of the data in bytes */
+    const unsigned char    *data;
+} OSSL_QUIC_FRAME_CRYPTO;
+
+/* QUIC Frame: RESET_STREAM */
+typedef struct ossl_quic_frame_reset_stream_st {
+    uint64_t    stream_id;
+    uint64_t    app_error_code;
+    uint64_t    final_size;
+} OSSL_QUIC_FRAME_RESET_STREAM;
+
+/* QUIC Frame: STOP_SENDING */
+typedef struct ossl_quic_frame_stop_sending_st {
+    uint64_t    stream_id;
+    uint64_t    app_error_code;
+} OSSL_QUIC_FRAME_STOP_SENDING;
+
+/* QUIC Frame: NEW_CONNECTION_ID */
+typedef struct ossl_quic_frame_new_conn_id_st {
+    uint64_t              seq_num;
+    uint64_t              retire_prior_to;
+    const unsigned char  *conn_id;
```

Comment:
It would be equally valid. Any preference?

Reply:
The ID cache proposal now uses fixed arrays.  We might as well be consistent and use them everywhere (and pray the standard doesn't increase the size soon).  Not that I'm against length + pointer.  The reduction in mallocs is probably a good thing.

BTW: I'm currently proposing an array 21 long, with the size in the first byte and the up to 20 octets in the remaining (zero filled).  That way one memcmp is enough to compare two.  It's a bit kludgy but workable.

Reply:
That's a bit obtuse. Since we will be representing connection IDs all over the place, I'll introduce a structure:
```
#define OSSL_QUIC_MAX_CONN_ID_LEN       20
typedef struct ossl_quic_conn_id_st {
    unsigned char   id_len;
    unsigned char   id[OSSL_QUIC_MAX_CONN_ID_LEN];
} OSSL_QUIC_CONN_ID;

```